ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -250,58 +255,77 @@ void close() throws IOException {
   /**
    * Checks the provided arguments for consistency and allocates the first segment file and creates the
    * {@link LogSegment} instance for it.
-   * @param totalCapacity the intended total capacity of the log.
    * @param segmentCapacity the intended capacity of each segment of the log.
+   * @return the {@link LogSegment} instance that is created.
    * @throws IOException if there is an I/O error creating the segment files or creating {@link LogSegment} instances.
    */
-  private void checkArgsAndAllocateFirstSegment(long totalCapacity, long segmentCapacity) throws IOException {
-    if (totalCapacity <= 0 || segmentCapacity <= 0) {
+  private LogSegment checkArgsAndGetFirstSegment(long segmentCapacity) throws IOException {
+    if (capacityInBytes <= 0 || segmentCapacity <= 0) {
       throw new IllegalArgumentException(
-          "One of totalCapacityInBytes [" + totalCapacity + "] or " + "segmentCapacityInBytes [" + segmentCapacity + "]"
-              + " is <=0");
+          "One of totalCapacityInBytes [" + capacityInBytes + "] or " + "segmentCapacityInBytes [" + segmentCapacity
+              + "] is <=0");
     }
-    segmentCapacity = Math.min(totalCapacity, segmentCapacity);
+    segmentCapacity = Math.min(capacityInBytes, segmentCapacity);
     // all segments should be the same size.
-    long numSegments = totalCapacity / segmentCapacity;
-    if (totalCapacity % segmentCapacity != 0) {
+    long numSegments = capacityInBytes / segmentCapacity;
+    if (capacityInBytes % segmentCapacity != 0) {
       throw new IllegalArgumentException(
-          "Capacity of log [" + totalCapacity + "] should be a multiple of segment capacity [" + segmentCapacity + "]");
+          "Capacity of log [" + capacityInBytes + "] should be a multiple of segment capacity [" + segmentCapacity
+              + "]");
     }
-    remainingUnallocatedSegments = numSegments;
-    String firstSegmentName = LogSegmentNameHelper.generateFirstSegmentName(numSegments);
-    logger.info("Allocating first segment with name [{}] and capacity {} bytes. Total number of segments is {}",
-        firstSegmentName, segmentCapacity, numSegments);
-    File segmentFile = allocate(LogSegmentNameHelper.nameToFilename(firstSegmentName), segmentCapacity);
+    Pair<String, String> segmentNameAndFilename = getNextSegmentNameAndFilename();
```

Comment:
Not necessarily - during compaction, the first target segment will have a specific name.